File: Client/Components/Inputs/InputBinary/InputBinary.razor.cs
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;

namespace FileFlows.Client.Components.Inputs;

/// <summary>
/// Input for a binary file
/// </summary>
public partial class InputBinary : Input<FileData>
{
    /// <summary>
    /// Gets or sets the JavaScript runtime
    /// </summary>
    [Inject] private IJSRuntime JS { get; set; }
    /// <summary>
    /// The JS Input Binary class instance
    /// </summary>
    private IJSObjectReference jsInputBinary;

    /// <summary>
    /// Translation strings
    /// </summary>
    private string lblBrowse;
    
    public override bool Focus() => FocusUid();
    async Task Browse()
    {
        var file = await jsInputBinary.InvokeAsync<FileData>("chooseFile");
        if (file == null)
            return;

        Value = file;

        // Notify Blazor the Value property has changed
        await ValueChanged.InvokeAsync(Value);
    }

    protected override async Task OnInitializedAsync()
    {
        await base.OnInitializedAsync();
        lblBrowse = Translater.Instant("Labels.Browse");
        var jsObjectReference = await jsRuntime.InvokeAsync<IJSObjectReference>("import", $"./Components/Inputs/InputBinary/InputBinary.razor.js?v={Globals.Version}");
        jsInputBinary = await jsObjectReference.InvokeAsync<IJSObjectReference>("createInputBinary");
    }
}


/// <summary>
/// File Data
/// </summary>
public class FileData
{
    /// <summary>
    /// Gets or sets the files mime type
    /// </summary>
    public string MimeType { get; set; }
    /// <summary>
    /// Gets or sets the files binary content
    /// </summary>
    public byte[] Content { get; set; }
}

File: Client/Components/Inputs/InputBinary/InputBinary.razor.js
/**
 * Creates the InputBinary instance and returns it
 * @returns {InputCode} the InputBinary instance
 */
export function createInputBinary()
{
    return new InputBinary();
}
/**
 * Class representing an InputBinary for handling file input.
 */
export class InputBinary 
{    
    /**
     * Opens a file dialog and retrieves the selected file's MIME type and content as a Uint8Array.
     * @returns {Promise<{mimeType: string, content: Uint8Array} | null>} An object with the MIME type and file content, or null if no file is selected.
     */
    async chooseFile() {
        return new Promise((resolve) => {
            const input = document.createElement("input");
            input.type = "file";
            input.onchange = async (e) => {
                const file = e.target.files[0];
                if (!file) {
                    resolve(null);
                    return;
                }

                const mimeType = file.type;
                const content = await file.arrayBuffer();
                resolve({mimeType, content: new Uint8Array(content)});
            };
            input.click();
        });
    }
}